--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -588,7 +588,7 @@ void br_fdb_update(struct net_bridge *br, struct net_bridge_port *source,
		/* attempt to update an entry for a local interface */
		if (unlikely(test_bit(BR_FDB_LOCAL, &fdb->flags))) {
			if (net_ratelimit())
-				br_warn(br, "received packet on %s with own address as source address (addr:%pM, vlan:%u)\n",
+				no_warn(br, "received packet on %s with own address as source address (addr:%pM, vlan:%u)\n",
					source->dev->name, addr, vid);
			unsigned long now = jiffies;
